maas-agent/src/charm.py

```diff
@@ -139,6 +140,30 @@ def _on_remove(self, _event: ops.RemoveEvent) -> None:
         except Exception as ex:
             logger.error(str(ex))
 
+    def _on_upgrade(self, _event: ops.UpgradeCharmEvent) -> None:
+        """Upgrade MAAS install on the machine.
+
+        Args:
+            event (ops.UpgradeCharmEvent): Event from ops framework
+        """
+        self.unit.status = ops.MaintenanceStatus(f"upgrading to {MAAS_SNAP_CHANNEL}...")
+        if current := MaasHelper.get_installed_channel():
+            if current > MAAS_SNAP_CHANNEL:
+                msg = f"Cannot downgrade {current} to {MAAS_SNAP_CHANNEL}"
+                # XXX: It looks like ErrorStatus is not valid in the version of Juju used by maas-charms
+                self.unit.status = ops.BlockedStatus(msg)
```

Comment:
Are we sure that setting the status here will not be overridden when `_on_collect_status` runs right after the `_on_upgrade` completion?

Reply:
Same for the region charm